crates/qasm2/src/parse.rs

```diff
@@ -0,0 +1,1796 @@
+    }
+
+    /// Parse a declaration of a classical register, emitting the relevant bytecode and adding the
+    /// definition to the relevant parts of the internal symbol tables in the parser state.  This
+    /// assumes that the `creg` token is still in the token stream.
+    fn parse_creg(&mut self, bc: &mut Vec<Option<InternalBytecode>>) -> PyResult<usize> {
+        let creg_token = self.expect_known(TokenType::Creg);
+        let name_token = self.expect(TokenType::Id, "a valid identifier", &creg_token)?;
+        let name = name_token.id(&self.context);
+        let lbracket_token = self.expect(TokenType::LBracket, "'['", &creg_token)?;
+        let size = self
+            .expect(TokenType::Integer, "an integer", &lbracket_token)?
+            .int(&self.context);
+        self.expect(TokenType::RBracket, "']'", &lbracket_token)?;
+        self.expect(TokenType::Semicolon, "';'", &creg_token)?;
+        match self.symbols.insert(
+            name.clone(),
+            GlobalSymbol::Creg {
+                size,
+                start: ClbitId::new(self.num_clbits),
+                index: CregId::new(self.num_cregs),
+            },
+        ) {
+            None | Some(GlobalSymbol::Gate { defined: false, .. }) => {
+                self.num_clbits += size;
+                self.num_cregs += 1;
+                bc.push(Some(InternalBytecode::DeclareCreg { name, size }));
+                Ok(1)
+            }
+            _ => Err(QASM2ParseError::new_err(message_generic(
+                Some(&Position::new(
+                    self.current_filename(),
+                    name_token.line,
+                    name_token.col,
+                )),
+                &format!("'{}' is already defined", name_token.id(&self.context)),
+            ))),
+        }
+    }
+
+    /// Parse a declaration of a quantum register, emitting the relevant bytecode and adding the
+    /// definition to the relevant parts of the internal symbol tables in the parser state.  This
+    /// assumes that the `qreg` token is still in the token stream.
+    fn parse_qreg(&mut self, bc: &mut Vec<Option<InternalBytecode>>) -> PyResult<usize> {
+        let qreg_token = self.expect_known(TokenType::Qreg);
+        let name_token = self.expect(TokenType::Id, "a valid identifier", &qreg_token)?;
+        let name = name_token.id(&self.context);
+        let lbracket_token = self.expect(TokenType::LBracket, "'['", &qreg_token)?;
+        let size = self
+            .expect(TokenType::Integer, "an integer", &lbracket_token)?
+            .int(&self.context);
+        self.expect(TokenType::RBracket, "']'", &lbracket_token)?;
+        self.expect(TokenType::Semicolon, "';'", &qreg_token)?;
+        match self.symbols.insert(
+            name.clone(),
+            GlobalSymbol::Qreg {
+                size,
+                start: QubitId::new(self.num_qubits),
+            },
+        ) {
+            None | Some(GlobalSymbol::Gate { defined: false, .. }) => {
```

Comment:
Can you give an example when this can match `Some(GlobalSymbol::Gate { defined: false, .. })`?

Reply:
I refactored a bunch of this code in 0816bfb00, so it should be rather clearer now hopefully.